includes/REST_API/Products_Controller.php

```diff
@@ -0,0 +1,489 @@
+ */
+
+namespace Google\Web_Stories\REST_API;
+
+use Google\Web_Stories\Infrastructure\HasRequirements;
+use Google\Web_Stories\Settings;
+use Google\Web_Stories\Story_Post_Type;
+use WP_Error;
+use WP_Http;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * Class to access publisher logos via the REST API.
+ *
+ * @since 1.12.0
+ *
+ * @phpstan-type ShopifyGraphQLError array{message: string}[]
+ * @phpstan-type ShopifyGraphQLPriceRange array{minVariantPrice: array{amount: int, currencyCode: string}}
+ * @phpstan-type ShopifyGraphQLProductImage array{url: string, altText: string}
+ * @phpstan-type ShopifyGraphQLProduct array{id: string, title: string, vendor: string, description: string, onlineStoreUrl: string, images: array{edges: array{node: ShopifyGraphQLProductImage}[]}, priceRange: ShopifyGraphQLPriceRange}
+ * @phpstan-type ShopifyGraphQLResponse array{errors?: ShopifyGraphQLError, data: array{products: array{edges: array{node: ShopifyGraphQLProduct}[]}}}
+ */
+class Products_Controller extends REST_Controller implements HasRequirements {
+
+	/**
+	 * Settings instance.
+	 *
+	 * @var Settings Settings instance.
+	 */
+	private $settings;
+
+	/**
+	 * Story_Post_Type instance.
+	 *
+	 * @var Story_Post_Type Story_Post_Type instance.
+	 */
+	private $story_post_type;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Settings        $settings Settings instance.
+	 * @param Story_Post_Type $story_post_type Story_Post_Type instance.
+	 */
+	public function __construct( Settings $settings, Story_Post_Type $story_post_type ) {
+		$this->settings        = $settings;
+		$this->story_post_type = $story_post_type;
+
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'products';
+	}
+
+	/**
+	 * Get the list of service IDs required for this service to be registered.
+	 *
+	 * Needed because the story post type needs to be registered first.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @return string[] List of required services.
+	 */
+	public static function get_requirements(): array {
+		return [ 'settings', 'story_post_type' ];
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes(): void {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_items' ],
+					'permission_callback' => [ $this, 'permissions_check' ],
+					'args'                => $this->get_collection_params(),
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to get and create items.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( ! $this->story_post_type->has_cap( 'edit_posts' ) ) {
+			return new \WP_Error(
```

Comment:
Why slash this and not use the class?

Reply:
We do it like this literally everywhere in the code base, IIRC it's a PHPCS config.